routing/localchans/manager.go

```diff
@@ -0,0 +1,201 @@
+package localchans
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/discovery"
+	"github.com/lightningnetwork/lnd/htlcswitch"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/routing"
+)
+
+// Manager manages the node's local channels. The only operation that is
+// currently implemented is updating forwarding policies.
+type Manager struct {
+	// UpdateForwardingPolicies is used by the manager to update active
+	// links with a new policy.
+	UpdateForwardingPolicies func(
+		chanPolicies map[wire.OutPoint]htlcswitch.ForwardingPolicy)
+
+	// PropagateChanPolicyUpdate is called to persist a new policy to disk
+	// and broadcast it to the network.
+	PropagateChanPolicyUpdate func(
+		edgesToUpdate []discovery.EdgeWithInfo) error
+
+	// ForAllOutgoingChannels is required to iterate over all our local
+	// channels.
+	ForAllOutgoingChannels func(cb func(*channeldb.ChannelEdgeInfo,
+		*channeldb.ChannelEdgePolicy) error) error
+
+	// FetchChannel is used to query local channel parameters.
+	FetchChannel func(chanPoint wire.OutPoint) (*channeldb.OpenChannel,
+		error)
+
+	// policyUpdateLock ensures that the database and the link do not fall
+	// out of sync if there are concurrent fee update calls. Without it,
+	// there is a chance that policy A updates the database, then policy B
+	// updates the database, then policy B updates the link, then policy A
+	// updates the link.
+	policyUpdateLock sync.Mutex
+}
+
+// UpdatePolicy updates the policy for the specified channels on disk and in the
+// active links.
+func (r *Manager) UpdatePolicy(newSchema routing.ChannelPolicy,
+	chanPoints ...wire.OutPoint) error {
+
+	r.policyUpdateLock.Lock()
+	defer r.policyUpdateLock.Unlock()
+
+	// First, we'll construct a set of all the channels that need to be
+	// updated.
+	chansToUpdate := make(map[wire.OutPoint]struct{})
+	for _, chanPoint := range chanPoints {
+		chansToUpdate[chanPoint] = struct{}{}
+	}
+
+	haveChanFilter := len(chansToUpdate) != 0
+
+	var edgesToUpdate []discovery.EdgeWithInfo
+	policiesToUpdate := make(map[wire.OutPoint]htlcswitch.ForwardingPolicy)
+
+	// Next, we'll loop over all the outgoing channels the router knows of.
+	// If we have a filter then we'll only collected those channels,
+	// otherwise we'll collect them all.
+	err := r.ForAllOutgoingChannels(func(
+		info *channeldb.ChannelEdgeInfo,
+		edge *channeldb.ChannelEdgePolicy) error {
+
+		// If we have a channel filter, and this channel isn't a part
+		// of it, then we'll skip it.
+		_, ok := chansToUpdate[info.ChannelPoint]
+		if !ok && haveChanFilter {
+			return nil
+		}
+
+		// Apply the new policy to the edge.
+		err := r.updateEdge(info.ChannelPoint, edge, newSchema)
```

Comment:
This is a bit strange, since the updated edge policy is invalid after this call (it hasn't been re-signed). A more natural split would be for the localChanManager to fully update and re-sign the policies, and then for the gossiper to just forward them around as any other network update. but maybe that's a change too big for this PR.

Reply:
Summary from offline discussion:
* Ideally gossiper would just gossip and not sign.
* In an `lnd` mode of operation where there is no gossip (all is private), there wouldn't be a need for signatures on local updates. Signing in `localchans.Manager` wouldn't be necessary either.
* An interconnecting component that takes policies from `localchans.Manager` and resigns them when they change or become stale could be a solution.

To keep the scope of this pr under control, I only added a call to clear the signature in `localchans.Manager`.